### DISCUSSION TITLE
Problem loading new fights generated with Hades Workshop

### DISCUSSION THREAD

**Discussion by Teamtorrentscripts** [Albeoris/Memoria#470] [Q&A] [answered]
Hello, I'm trying to generate new battles with HW, but I'm having a problem loading them into the game.

I did a test by re-editing a combat from the game, but adding it to another new combat with ID5000, so far so good, but when the game is loaded, it keeps flashing continuously, without loading anything else. I've tried changing by putting a new ID, but it's still the same. Does anyone know how to fix it?

Thanks in advance

![image](https://github.com/Albeoris/Memoria/assets/169691119/00087c46-3247-4fde-bf77-869ff0c45ecf)

https://github.com/Albeoris/Memoria/assets/169691119/221036f0-747f-4410-86e3-f12c71f2169a

**Comment by Tirlititi** [comment, marked as answer]
I have no problem with that from my end.
![CustomBattleTriggerer](https://github.com/Albeoris/Memoria/assets/20627826/6592e7b1-290e-4094-b300-40ef39b8aa40)
![CustomBattleExport](https://github.com/Albeoris/Memoria/assets/20627826/e2465cc3-2c01-43cb-9f9e-38910a4e8488)
![CustomBattleInGame](https://github.com/Albeoris/Memoria/assets/20627826/6bb19fb1-ec0f-4ab6-ac56-c2f077ff928f)

That looks more like you have a mismatch between the number of enemies you declare in the group ("Group 1" -> "Enemy Amount", or their types) and the number of enemies you setup in the AI script ("Edit Script" -> "Function Main_Init" -> the number of lines "InitObject" with proper parameters).

**Reply by Teamtorrentscripts** [reply]
Well, it is very strange because the new battle exported was without touching anything of it, I chose the base battle 280 with ID 5000 again as it was (without modifying anything about that same battle) and the same thing happened again. Could it be that I put the files wrong when the battle is exported? This is how I leave them in my mod.

Please help :(

![image](https://github.com/Albeoris/Memoria/assets/169691119/79d59d78-01f5-442f-88e3-aface3d6e28f)
![image](https://github.com/Albeoris/Memoria/assets/169691119/48f708c7-1c0c-458f-b875-616b8adbc093)
![image](https://github.com/Albeoris/Memoria/assets/169691119/cb50f57b-83dc-4698-93da-7ae3b1ec4483)
![image](https://github.com/Albeoris/Memoria/assets/169691119/b5304b07-159e-4658-b7c3-46ddbb2ba809)
![image](https://github.com/Albeoris/Memoria/assets/169691119/0503ee60-5d40-4fa0-8008-3e5fb2cd4872)

**Reply by Tirlititi** [reply]
What's that line ``BattleMapModel GEO_WEP_B1_096`` in the dictionary patch? It's not correct.
That shouldn't be the cause of the crash though; incorrect dictionary patch lines are just ignored, unless I'm mistaken.

The rest of what you showed seem correct. Do you have the following files as well?
```txt
Rikkuandyuna Edition/StreamingAssets/assets/resources/BattleMap/BattleScene/EVT_BATTLE_CUSTOM_BATTLE_000/5000.raw17.bytes
Rikkuandyuna Edition/StreamingAssets/assets/resources/BattleMap/BattleScene/EVT_BATTLE_CUSTOM_BATTLE_000/dbfile0000.raw16.bytes
Rikkuandyuna Edition/FF9_Data/embeddedasset/text/es/Battle/5000.mes
```
Also, you don't show how you trigger that battle in-game. Do you modify the script of a field? the script of a world map? the battle sets of a world map?

**Reply by Teamtorrentscripts** [reply]
The "BattleMapModel GEO_WEP_B1_096" in DiccionaryPatch I seem to remember that it was already there, but it had never given me any error of any kind before

I also already found the error according to your answer! I needed to put the text file!! I put it in and it worked the first time. You handle so many files that you sometimes forget to put them in hahaha.

Thank you very much, you always solve my problems!!

**Comment by Teamtorrentscripts** [comment]
One question, to export a new field with a new ID, where are the files generated by HW placed in the mod? There if I get lost and I can't find them.
![image](https://github.com/Albeoris/Memoria/assets/169691119/36768688-1ed6-4355-a377-99f8afb78e92)

**Reply by Tirlititi** [reply]
You have the different asset paths that are required for a field (and thus the files that are generated or copied by HW for these custom fields) listed [here](https://github.com/Albeoris/Memoria/wiki/Dictionary-Patch#user-content-FieldScene).

So basically:
```txt
Assets/Resources/FieldMaps/[Field Map Identifier]/atlas.png
Assets/Resources/FieldMaps/[Field Map Identifier]/[Field Map Identifier].bgi.bytes
Assets/Resources/FieldMaps/[Field Map Identifier]/[Field Map Identifier].bgs.bytes
Assets/Resources/CommonAsset/EventEngine/EventBinary/Field/[LANG]/[Script Identifier].eb.bytes
```

The files placed in "FF9_Data" are mostly the files containing texts (there are a couple of other things in "FF9_Data" but the majority is text). For fields, you don't necessarily need a custom text block, so there's none by default. If you do want to separate the texts and dialogs relevant to your custom fields, you can use a [MessageFile](https://github.com/Albeoris/Memoria/wiki/Dictionary-Patch#user-content-MessageFile) line in your Dictionary Patch.

**Reply by Teamtorrentscripts** [reply]
Well, I must do something wrong, when loading that field ID in the game, the screen appears like this
![image](https://github.com/Albeoris/Memoria/assets/169691119/e2211017-8c70-4163-a5b7-e33aa365b2f9)

I put all the files in the location he told me and I don't know what could go wrong :S. This is how I load that Field ID with HW.

![image](https://github.com/Albeoris/Memoria/assets/169691119/ff31de83-9f3a-4da5-a72b-6007e54fd155)

**Reply by Tirlititi** [reply]
Ah, that's because HW uses the default non-modded ``atlas.png`` of the field. But since you are playing with the Moguri Mod, the game expects field backgrounds to be upscaled. You must extract the related ``atlas.png`` of the field from the ``MoguriFiles`` archives instead.

It's hacky, but the way to do it is to make backups of your files ``FINAL FANTASY IX/StreamingAssets/p0data1X.bin``, then copy-paste (or move) the version of these files you'll find inside the ``MoguriFiles`` folder, then open HW and "Tools -> Unity Assets Viewer -> Open -> FF9_Launcher.exe" then "Archives -> p0data1" and then browse these archives to find the atlas [corresponding to the field](https://github.com/Albeoris/Memoria/wiki/List-of-fields) you're based on. In this case, that's Prima Vista's Cargo Room so it's the following asset:
```txt
Assets/Resources/FieldMaps/FBG_N00_TSHP_MAP002_TH_CGR_1/atlas.png
```
Then "right-click -> Export Selection" and you'll have that asset in a sub-folder of "FINAL FANTASY IX/HadesWorkshopAssets".
Move that asset in place of the ``atlas.png`` previously exported by HW in your mod folder.

And then, move the archives p0data1 back to their Moguri mod folder and replace the backup of the original archives.

Most likely, you'd want to export all the ``atlas.png`` you think you'll use before putting the backups back to their original place, because you don't want to do that procedure everytime. Also keep a backup of these atlas you export somewhere, in case they get replaced when exporting custom fields with HW.

**Reply by Teamtorrentscripts** [reply]
I forgot to say that I used Mogurimod, sorry ;(

But with her instructions it has been easier than I thought! Everything has worked without problems, I load that same map with the new ID without any errors.

I am happy thanks to you!! thank you very much once again ;)

**Comment by Teamtorrentscripts** [comment]
Excuse me, how was it to import the new dialogs that I incorporated into the new field ID? I'm trying this in Diccionari Patch but it doesn't work, it loads other different dialogs, the new dialogs are the 3 highlighted in red

<img width="1273" alt="Captura de pantalla 2024-05-17 221328" src="https://github.com/Albeoris/Memoria/assets/169691119/e4f01bd9-f34a-4b7d-9fe8-3d76f7cd2d5c">

![image](https://github.com/Albeoris/Memoria/assets/169691119/e1771be8-fc3e-40b6-81e0-90525f3da89f)

![image](https://github.com/Albeoris/Memoria/assets/169691119/2b4b3013-4ad8-4d4d-afaa-26bb3384d4a7)

**Reply by Tirlititi** [reply]
If it's only a couple of lines, you don't need to have a new text block or a Dictionary Patch line for that. You only need to edit the field's script. Dialogs in cutscenes are displayed by lines ``WindowSync``, ``WindowAsync`` and a couple other codes with almost the same name. Try to find these lines in the field's script.

**Reply by Teamtorrentscripts** [reply]
Hello again, I tried to do what you told me, but I can't solve the problem, I add new texts and implement them in the corresponding field id, but when loading the new ID generated by HW, the dialogs do not correspond to those that I put

Did I do something wrong? I have it like this (the new dialogues are those indicated in the image)

![image](https://github.com/Albeoris/Memoria/assets/169691119/fd5dac20-5c0a-47d2-a46a-3e46fcdafcad)

**Reply by Tirlititi** [reply]
Sorry I don't understand what exactly is the problem.
You have everything working fine in HW but it doesn't work in-game? Or is there a problem in HW already?
What's the result that you get? What did you expect instead?

**Reply by Teamtorrentscripts** [reply]
Let me explain, what I do is simply copy a field ID and then export it with another different ID with HW (as with combats)

I do the process and it is done well, when I put Field 4000 (which is the new configured) in a field on an NPC, the map loads correctly.

The problem comes when I want to put new messages in that field about the existing NPCs of that same field ID (4000), but when I do so, the ones I chose in HW do not load (I put them through the texts tab, and in HW if it works correctly for me)

The ones that load me in the game are different dialogues, and that is what I want to fix.

**Reply by Tirlititi** [reply]
In [your screenshots](https://github.com/Albeoris/Memoria/discussions/470#discussioncomment-9474842), you use the text block "Treno (1)" in some places and "Treno (2)" in other places. Make sure that you use a consistant text block:
- The first screenshot of the message uses "Treno (1)"
- The second screenshot of the message uses "Treno (2)"
- The third screenshot setups a whole new text block, which is not required

**Reply by Teamtorrentscripts** [reply]
Ok ok, I already found the bug, apparently it was not saved correctly in HW, a very strange thing but oh well :S.

Now I wanted to ask you (sorry if it's a different topic)

I am looking for a way to modify an existing combat in the game, and have said combat appear with low frequency in the Gizamaluke area. As I have seen, this can be done by adjusting the "Frequency" option in the group section within the enemy, right?

The problem I have with that is that when I lower it, for example to 25, when the combat appears in the game, it freezes and I don't understand why :(. Here is the HW screenshot.

Thank you very much for all the help and your time!

![image](https://github.com/Albeoris/Memoria/assets/169691119/24a62f08-4afa-46c2-afc1-270c01838080)

**Reply by Tirlititi** [reply]
No, there are multiple steps when deciding which enemies will be fought in a random battle. Thus there are 2 different frequencies and you mix them up (it's relatively normal though as it's a bit tricky and not explained in evidence in the tool):

1) First step: when a random encounter triggers, the battle that is fought is picked amongst a pool of 4 battles. On the World Maps, these are relatively easy to find in the panel "Environment -> World Map -> Battle Spots". However, on the fields, they are defined in the field script. It's the lines ``SetRandomBattles`` and they are almost always in either the function ``Main_Init`` or ``Main_Loop``.

2) Second step: once a battle is picked, there are still variations of the same battle that can be used, typically with a different number of enemies. These are the different battle groups. You must always have the last group's frequency set at 255 (I should have greyed out the parameter for the last group in order to prevent people from changing that, sorry). It's a 100-based chance and the checks are done in the group order, so a frequency of "40" means "40% chances to use this group, provided that the battle was picked in step 1, unless the sum of frequencies of the previous groups already exceed 100%". That is why the last group is always at 255: it's a fallback that will always be picked if the other groups were not picked before.

You want to change the 1st step. Unfortunatly, it's preset settings, meaning that you cannot set whichever frequency you want there, but only one of the different frequency distributions that are possible.

**Reply by Teamtorrentscripts** [reply]
Ohh I see, it seems confusing, I'll put this aside then and focus on other things I'm trying.

I also wanted to ask, I have this set in the ATB section of an enemy (grand dragon), is it set correctly? This attack is supposed to be used with a 25% chance, but I see him use it almost every time.

![image](https://github.com/Albeoris/Memoria/assets/169691119/29df3f62-a504-4173-819c-2f46426e6608)

**Reply by Tirlititi** [reply]
``(GetRandom % 4) == 0`` instead and you're good.

``GetRandom % 4`` returns a number between 0 and 3, and the condition is fullfilled whenever the number is anything except 0.

**Reply by Teamtorrentscripts** [reply]
I don't know how to thank you for all the help, I tried it now and it works fine.

For now I have no more doubts, if something arises I will turn to your help. Thanks a lot ;)

**Reply by Teamtorrentscripts** [reply]
Hello again, I have another question: how can I make it so that in battle, when the enemy casts a spell, a text appears afterwards?


I know that the texts in combat are called "BattleDialog" in the scripts of each enemy, but I can't get it to work after they cast any spell

![image](https://github.com/Albeoris/Memoria/assets/169691119/d7a0806d-da6d-43ba-8e4f-fc809caa5db6)

**Reply by Tirlititi** [reply]
It is possible to do it with ``BattleDialog`` lines in scripts (you'd do it inside the "Loop" function instead), but it's better and easier, for what I understand you want to do, to add it to the enemy's ability sequence. On the main frame, select the enemy's spell / ability for which you want to add a message and click on "Edit Animation".
There, click on a small "+" sign to add a new line near the end of the sequence and use either "Battle Text" or "Battle Title" to add a message. You may need to add / tweak lines "Wait" in order to setup the exact moment you want this message to appear during the enemy's move.

**Reply by Teamtorrentscripts** [reply]
I could swear that I had tried that a long time ago without it working well... maybe I did it wrong at the time, because in this attempt it did work hahaha.

And another thing, so that the enemies execute the attacks in order according to how they are placed in their ATB? That is, according to this capture, the attacks are executed in this order: 1, 2, 3, 4, 5, 6 and 7, and then it starts again.

I say this more than anything because for me the enemy makes skipped attacks (first 1, then 6, etc.)

![image](https://github.com/Albeoris/Memoria/assets/169691119/893c4d6d-38e4-4bd5-b4d7-25bba650d5e6)

**Reply by Tirlititi** [reply]
Can't say what's happening without seeing the whole function. Please copy-paste it instead.

**Reply by Teamtorrentscripts** [reply]
ohh, sorry, I leave it here

Function Mímico_dorado_ATB
    set #( SV_Target = RandomInTeam(( NotMatching(SV_PlayerTeam[STATUS_CURRENT_A], 257) & NotMatching(SV_PlayerTeam[STATUS_CURRENT_B], 64) )) )
    switchex 8 ( GetRandom % 5 ) {
    case 0:
        Attack( 0 )
        break
    case 1:
        Attack( 6 )
        break
    case 2:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 8 )
    case 3:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 5 )
    case 4:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 3 )
    case 5:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 4 )
    case 6:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 7 )
    case 7:
        if ( ( GetRandom % 60 ) == 0 ) {
            set #( SV_Target = SV_PlayerTeam )
            Attack( 10 )
        }
    }
    return

**Reply by Tirlititi** [reply]
That's not doing the attacks one after the other, that's doing them randomly (and not all of them, only the first 5 cases).

"``switchex 8 ( GetRandom % 5 )``" introduces 8 different code blocks (called ``case``) with only one of them being executed, depending on the result of the formula in parenthesis. There, the formula is ``GetRandom % 5``, which means that everytime the function ATB runs, a random number between 0 and 4 is picked and the corresponding ``case`` is executed.

In order to have the different attacks being used one after the other, you rather need to use a variable that keeps track of the next attack to be used. You do that with these lines inside the "Local variables" block above the function's code:
```txt
allocate 1
local uint8 attack_counter
```
And then, you use that variable in the ``switchex`` formula:
```txt
switchex 8 ( attack_counter ) {
case 0:
    // First attack....
    break
case 1:
    // Second attack...
    break
// etc...
}
// Increment the attack counter
set attack_counter++
// Make it loop back to the first attack if it got too high
if ( attack_counter >= 8 ) {
    set attack_counter = 0
}
return
```
Don't forget the ``break`` keyword at the end of each case!

**Reply by Teamtorrentscripts** [reply]
Ugh, I think I got lost in your answer haha, I'm trying to see if I understood it, but after more than 10 minutes trying, HW never saves the applied changes.

If you would be so kind, could you put the complete code so that these attacks occur in order? so I'm getting ideas

Thank you so much for your patience ;)

 set #( SV_Target = RandomInTeam(( NotMatching(SV_PlayerTeam[STATUS_CURRENT_A], 257) & NotMatching(SV_PlayerTeam[STATUS_CURRENT_B], 64) )) )
    switchex 3 ( GetRandom % 3 ) {
    case 0:
        Attack( 0 )
        break
    case 1:
        Attack( 1 )
        break
    case 2:
        set #( SV_Target = SV_PlayerTeam )
        Attack( 7 )
        break
        }
    return

**Reply by Teamtorrentscripts** [reply]
Sorry to bother you again @Tirlititi , I managed to get the previous problem to work, but now I find that when I put a new attack on the enemy (attack 4 in the capture), it does it randomly and not in order.

Besides, I also notice that when an ally attacks the enemy, he only focuses on him until he kills him, and I don't want that

Do you know why it is? I have it like this in HW

![image](https://github.com/Albeoris/Memoria/assets/169691119/f4d2f888-56dd-4ab5-8fcd-8f08beff2725)

**Reply by Tirlititi** [reply]
You need to setup a target before each ``Attack`` line. The standard targets are either ``SV_PlayerTeam`` for an enemy's multi-target attack, and ``RandomInTeam(SV_PlayerTeam)`` for single-target attacks.

So let's say your attacks 0 and 1 are multi-target, your attacks 2 and 3 are single-target, and let's say your attack 4 is single-target and should target in priority a character that has Trouble:
```txt
switchex 5 ( attack_counter ) {
case 0:
    set SV_Target = SV_PlayerTeam
    Attack( 0 )
    break
case 1:
    set SV_Target = SV_PlayerTeam
    Attack( 1 )
    break
case 2:
    set SV_Target = RandomInTeam(SV_PlayerTeam)
    Attack( 2 )
    break
case 3:
    set SV_Target = RandomInTeam(SV_PlayerTeam)
    Attack( 3 )
    break
case 4:
    set SV_Target = RandomInTeam( Matching(SV_PlayerTeam[STATUS_CURRENT_A], 32) )
    if ( SV_Target == 0 ) {
        set SV_Target = RandomInTeam(SV_PlayerTeam)
    }
    Attack( 4 )
    break
}
set attack_counter++
if ( attack_counter >= 5 ) {
    set attack_counter = 0
}
return
```

**Reply by Teamtorrentscripts** [reply]
Wow, I get it, with that code it seems to be working fine and there are no problems. Thank you.

Now I'm looking for an attack that only executes when I do another attack a certain number of times. Example; When you do Attack (4) 3 times, immediately cast Attack 1 every time (or else cast Attack 1 and then reset everything)
